pkg/broker/cel.go

```diff
@@ -0,0 +1,137 @@
+	// vars so extensions can be future proofed.
+)
+
+func (r *Receiver) filterEventByCEL(ts *eventingv1alpha1.TriggerSpec, event *cloudevents.Event) (bool, error) {
+	e, err := cel.NewEnv(
+		cel.Types(&celprotos.CloudEventContext{}),
+		cel.Declarations(
+			decls.NewIdent(CELVarKeyContext, decls.NewObjectType("dev.knative.CloudEventContext"), nil),
+			decls.NewIdent(CELVarKeyExtensions, decls.NewObjectType("google.protobuf.Struct"), nil),
+			decls.NewIdent(CELVarKeyData, decls.NewObjectType("google.protobuf.Struct"), nil),
+		),
+	)
+	if err != nil {
+		return false, err
+	}
+
+	p, iss := e.Parse(ts.Filter.CEL.Expression)
+	if iss != nil && iss.Err() != nil {
+		return false, iss.Err()
+	}
+	c, iss := e.Check(p)
+	if iss != nil && iss.Err() != nil {
+		return false, iss.Err()
+	}
+
+	//TODO cache these by hash of expression. Programs are thread-safe so it's
+	// ok to share them between triggers and events.
+	prg, err := e.Program(c)
+	if err != nil {
+		return false, err
+	}
+
+	vars := map[string]interface{}{}
+	// Set baseline context fields
+	ceCtx := &celprotos.CloudEventContext{
+		Specversion:     event.Context.GetSpecVersion(),
+		Type:            event.Context.GetType(),
+		Source:          event.Context.GetSource(),
+		Schemaurl:       event.Context.GetSchemaURL(),
+		Datamediatype:   event.Context.GetDataMediaType(),
+		Datacontenttype: event.Context.GetDataContentType(),
+	}
+	vars[CELVarKeyContext] = ceCtx
+
+	// If the Trigger has requested parsing of extensions, attempt to turn them
+	// into a dynamic struct.
+	if ts.Filter.CEL.ParseExtensions {
+		//TODO should this coerce to V02?
+		extStruct, err := ceParsedExtensionsStruct(event.Context.AsV02().Extensions)
+		if err != nil {
+			r.logger.Error("Failed to parse event context for CEL filtering", zap.String("id", event.Context.AsV02().ID), zap.Error(err))
+		} else {
+			vars[CELVarKeyExtensions] = extStruct
+		}
+	}
+
+	// If the Trigger has requested parsing of data, attempt to turn them into
+	// a dynamic struct.
+	if ts.Filter.CEL.ParseData {
+		dataStruct, err := ceParsedDataStruct(event)
+		if err != nil {
+			r.logger.Error("Failed to parse event data for CEL filtering", zap.String("id", event.Context.AsV02().ID), zap.Error(err))
+		} else {
+			vars[CELVarKeyData] = dataStruct
+		}
+	}
+
+	out, _, err := prg.Eval(vars)
+	if err != nil {
+		return false, err
+	}
+
+	return out == types.True, nil
+}
+
+func ceParsedExtensionsStruct(ext map[string]interface{}) (*structpb.Struct, error) {
+	extJSON, err := json.Marshal(ext)
+	if err != nil {
+		return nil, err
+	}
+
+	extStruct := &structpb.Struct{}
+	if err := jsonpb.Unmarshal(bytes.NewBuffer(extJSON), extStruct); err != nil {
+		return nil, err
+	}
+	return extStruct, nil
+}
+
+func ceParsedDataStruct(event *cloudevents.Event) (*structpb.Struct, error) {
+	//TODO CloudEvents SDK might have a better way to do this with data codecs
+	if event.Context.GetDataContentType() == "application/json" {
```

Comment:
Use `GetMediaType` here.

Reply:
Did you just review your own code? :stuck_out_tongue_closed_eyes:

Reply:
I was pair-reviewing with Scott :smile: